### PR TITLE
[libcxx][libcxxabi] Fix build for OpenBSD

### PR DESCRIPTION
**Pull request by Ericson2314**
- No indirect syscalls on OpenBSD. Instead there is a `futex` function which issues a direct syscall.

- monotonic clock is avaialable despite the full POSIX suite of timers not being available in its entirety.

  See https://lists.boost.org/boost-bugs/2015/07/41690.php and https://github.com/boostorg/log/commit/c98b1f459add14d5ce3e9e63e2469064601d7f71 for a description of an analogous problem and fix for Boost.

### CODE REVIEW

**Review by brad0 — APPROVED** [review]

**Comment by brad0** [comment]
@Ericson2314 I was looking around regarding this and I also noticed src/filesystem/filesystem_clock.cpp and CLOCK_GETTIME.

**Comment by ldionne** [comment]
Folks, please don't merge without getting an approval from the libc++/libc++abi reviewers in the future. This is the policy we follow in libc++. I've been OOO for the past 10 days so please let me know if somehow this followed a different approval path and I just missed it, but on the surface it looks like this change didn't go through our normal approval process.

About the patch itself: Is there a reason why this is a problem now but wasn't a problem before? Also, we claim to support FreeBSD but we don't claim to support OpenBSD at all, and in order to do that we'd need to have a CI tester running per our policy.

**Comment by Ericson2314** [comment]
Sorry @ldionne. I didn't have so much of an idea of who is or isn't a reviewer, so didn't know anything was proceeding out of the ordinarily.

> About the patch itself: Is there a reason why this is a problem now but wasn't a problem before?

I imagine no one tried to build it on OpenBSD before.

**Comment by ldionne** [comment]
Screenshot 2024-05-22 at 10 49 38

This shows what approvals are necessary before landing a patch. The review group must have approved.

> I imagine no one tried to build it on OpenBSD before.

Ok, so then this patch is adding support for a new platform. We have a policy for doing that explained in https://libcxx.llvm.org/#platform-and-compiler-support. As part of our policy, we require CI to be present for all platforms that we officially support.

Since this is not the case right now and it seems that more patches are coming down the pipeline to support OpenBSD, I would like us to add such support before we proceed with any other OpenBSD support patch. If we can't fulfill our support policy, I would ask to revert this patch.

It's unfortunate that this was cherry-picked to the LLVM 18 release since this is definitely not something I'd consider a bugfix, but we can live with that.

**Comment by brad0** [comment]
> About the patch itself: Is there a reason why this is a problem now but wasn't a problem before? Also, we claim to support FreeBSD but we don't claim to support OpenBSD at all, and in order to do that we'd need to have a CI tester running per our policy.

OpenBSD very recently made a change removing syscall(2) otherwise it's built fine for many many years.

**Comment by brad0** [comment]
> I imagine no one tried to build it on OpenBSD before.

It's our system compiler / C++ standard library..

**Comment by brad0** [comment]
> > I imagine no one tried to build it on OpenBSD before.
> 
> Ok, so then this patch is adding support for a new platform. We have a policy for doing that explained in https://libcxx.llvm.org/#platform-and-compiler-support. As part of our policy, we require CI to be present for all platforms that we officially support.

It's not a new platform per se. OpenBSD made a local change that broke things. We've been building and using libc++ for many years.
 
> Since this is not the case right now and it seems that more patches are coming down the pipeline to support OpenBSD, I would like us to add such support before we proceed with any other OpenBSD support patch. If we can't fulfill our support policy, I would ask to revert this patch.

There are no other patches coming down the pipeline. There is nothing else at all. The one local patch we had for libcxxabi was commited as part of this patch.

**Comment by ldionne** [comment]
I think I saw https://github.com/llvm/llvm-project/pull/92675 which is also for OpenBSD. Generally speaking, what I'm saying is that we don't officially support platforms for which we have no tests, and OpenBSD currently falls in that category. I think it's great that OpenBSD uses libc++ as its system library and I would like the OpenBSD folks to work on getting us official test coverage on that platform so we can support it as a first class citizen.

**Review by Mhunter15 — COMMENTED** [review]
#94840

**Pull request merged**
